src/BuiltInTools/dotnet-watch/EnvironmentVariablesBuilder.cs

```diff
@@ -3,51 +3,99 @@
 
 namespace Microsoft.DotNet.Watcher
 {
-    internal sealed class EnvironmentVariablesBuilder : Dictionary<string, string>
+    internal sealed class EnvironmentVariablesBuilder
     {
         private static readonly char s_startupHooksSeparator = Path.PathSeparator;
         private const char AssembliesSeparator = ';';
 
-        public List<string> DotNetStartupHooks { get; } = [];
-        public List<string> AspNetCoreHostingStartupAssemblies { get; } = [];
+        public List<string> DotNetStartupHookDirective { get; } = [];
+        public List<string> AspNetCoreHostingStartupAssembliesVariable { get; } = [];
+
+        /// <summary>
+        /// Environment variables set on the dotnet run process.
+        /// </summary>
+        private readonly Dictionary<string, string> _variables = [];
+
+        /// <summary>
+        /// Environment variables passed as directives on command line (dotnet [env:name=value] run).
+        /// Currently, the effect is the same as setting <see cref="_variables"/> due to
+        /// https://github.com/dotnet/sdk/issues/40484
+        /// </summary>
+        private readonly Dictionary<string, string> _directives = [];
 
         public static EnvironmentVariablesBuilder FromCurrentEnvironment()
         {
             var builder = new EnvironmentVariablesBuilder();
 
             if (Environment.GetEnvironmentVariable(EnvironmentVariables.Names.DotnetStartupHooks) is { } dotnetStartupHooks)
             {
-                builder.DotNetStartupHooks.AddRange(dotnetStartupHooks.Split(s_startupHooksSeparator));
+                builder.DotNetStartupHookDirective.AddRange(dotnetStartupHooks.Split(s_startupHooksSeparator));
             }
 
             if (Environment.GetEnvironmentVariable(EnvironmentVariables.Names.AspNetCoreHostingStartupAssemblies) is { } assemblies)
             {
-                builder.DotNetStartupHooks.AddRange(assemblies.Split(AssembliesSeparator));
+                builder.AspNetCoreHostingStartupAssembliesVariable.AddRange(assemblies.Split(AssembliesSeparator));
             }
 
             return builder;
         }
 
-        public void AddToEnvironment(Dictionary<string, string> variables)
+        public void SetDirective(string name, string value)
         {
-            foreach (var (name, value) in this)
+            // Values that need to be aggregated:
+            if (name.Equals(EnvironmentVariables.Names.DotnetStartupHooks, StringComparison.OrdinalIgnoreCase))
```

Comment:
Is there any conceivable way for SetDirective to be called multiple times with the "DotnetStartupHooks" as the directive name?  I'm considering whether or not the .Add below might throw.

Reply:
I'm gonna refactor this a bit.